src/MassTransit/Transports/ConsumerAgent.cs

```diff
@@ -174,9 +189,22 @@ protected virtual async Task ActiveAndActualAgentsCompleted(StopContext context)
             }
         }
 
-        protected Task Dispatch(ReceiveContext context, ReceiveLockContext receiveLock = default)
+        protected async Task Dispatch<TContext>(TKey key, TContext context, ReceiveLockContext receiveLock = default)
+            where TContext : BaseReceiveContext
         {
-            return _dispatcher.Dispatch(context, receiveLock);
+            var added = _pending.TryAdd(key, context);
```

Comment:
Here is an interesting concept. If the message is already pending, it's likely a redelivery from the transport. Would be interesting to "hold" that message until the pending one completes, and then come up with a way to see if it's still "locked" or not and not deliver it. Just a thought.

Reply:
It is probably a next step, and should be NOTED :)